From 1d3ea526ce3be6afcf009c338c2500fb1a1b28fc Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Tue, 10 Sep 2024 07:23:38 -0700
Subject: [PATCH] winex11: Force disable glXWaitForSbcOML (AMD) on Wayland to
 circumvent Xwayland locking FPS to the monitor's refresh rate.

It should only break ingame vsync, but WINE_XWAYLAND_FPS_LOCK_WORKAROUND=0 can override-disable this behavior.
---
 dlls/winex11.drv/opengl.c |  3 ++-
 dlls/winex11.drv/x11drv.h | 12 ++++++++++++
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/opengl.c b/dlls/winex11.drv/opengl.c
index b6c73b8d0af..aaa9956808d 100644
--- a/dlls/winex11.drv/opengl.c
+++ b/dlls/winex11.drv/opengl.c
@@ -714,7 +714,8 @@ static void init_opengl(void)
 
     if (has_extension( glxExtensions, "GLX_OML_sync_control" ))
     {
-        pglXWaitForSbcOML = pglXGetProcAddressARB( (const GLubyte *)"glXWaitForSbcOML" );
+        if (!disable_glXWaitForSbcOML())
+            pglXWaitForSbcOML = pglXGetProcAddressARB( (const GLubyte *)"glXWaitForSbcOML" );
         pglXSwapBuffersMscOML = pglXGetProcAddressARB( (const GLubyte *)"glXSwapBuffersMscOML" );
     }
 
diff --git a/dlls/winex11.drv/x11drv.h b/dlls/winex11.drv/x11drv.h
index 00c54b19767..d967c721499 100644
--- a/dlls/winex11.drv/x11drv.h
+++ b/dlls/winex11.drv/x11drv.h
@@ -944,7 +944,19 @@ static inline BOOL intersect_rect( RECT *dst, const RECT *src1, const RECT *src2
     dst->bottom = min( src1->bottom, src2->bottom );
     return !IsRectEmpty( dst );
 }
+static inline BOOL disable_glXWaitForSbcOML(void)
+{
+    const char *override_env = getenv("vblank_mode");
+    if (override_env)
+        if (*override_env != '\0' && *override_env != '0')
+            return FALSE;
+        else
+            return TRUE;
+
+    if (!getenv("WAYLAND_DISPLAY")) return FALSE;
 
+    return TRUE;
+}
 /* registry helpers */
 
 extern HKEY open_hkcu_key( const char *name );
-- 
2.46.0

